ocaml/typing/typecore.ml

```diff
@@ -3428,58 +3458,93 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 mode_fun sargs ret
                List.map
-                 (fun (marg, exp) ->
-                    submode ~loc:exp.exp_loc ~env ~reason:Other
+                 (fun (marg, loc) ->
+                    submode ~loc ~env ~reason:Other
                       marg (mode_partial_application expected_mode);
                     Value.regional_to_local_alloc marg)
                  open_args
              in
              let closed_args = new_closed_args @ closed_args in
              let open_args = [] in
-             let mode_closed_args = List.map Alloc.close_over closed_args in
-             let mode_partial_fun = Alloc.partial_apply mode_fun in
+             (* note that [mode_fun] is closed_over, not partial_apply *)
+             let all_closed = mode_fun :: closed_args @ underscore_args in
              let mode_closure, _ =
-               Alloc.newvar_above (Alloc.join
-                (mode_partial_fun:: mode_closed_args))
+               all_closed |> List.map Alloc.close_over |> Alloc.join |> Alloc.newvar_above
              in
              register_allocation_mode mode_closure;
              let arg = Omitted { mode_closure; mode_arg; mode_ret; sort_arg } in
              let args = (lbl, arg) :: args in
-             (ty_ret, mode_closure, open_args, closed_args, args))
-      (ty_ret, mode_ret, [], [], []) (List.rev args)
+             (ty_ret, mode_closure, open_args, closed_args, underscore_args, args))
+      (ty_ret, mode_ret, [], [], [], []) (List.rev args)
+  in
+  let closed_args =
+    if List.length underscore_args = 0 then closed_args
+    else
+    let new_closed_args =
+      List.map
+        (fun (marg, loc) ->
+          submode ~loc ~env ~reason:Other
+            marg (mode_partial_application expected_mode);
+          Value.regional_to_local_alloc marg)
+        open_args
+    in
+    funct_mode :: new_closed_args @ closed_args
+  in
+  let rec type_underscore_parameters closed_args = function
+    | [] -> ty_ret, mode_ret, []
+    | (lbl, Dummy { mode_fun = _; ty_arg; mode_arg; level; sort_arg }) :: rest ->
+        let ty_ret, mode_ret, args = type_underscore_parameters (mode_arg :: closed_args) rest in
+        let arrow_desc = (lbl, mode_arg, mode_ret) in
+        let ty_ret =
+          newty2 ~level
+            (Tarrow (arrow_desc, ty_arg, ty_ret, commu_ok))
+        in
+        let mode_closure, _ =
+          closed_args |> List.map Alloc.close_over |> Alloc.join |> Alloc.newvar_above
+        in
+        register_allocation_mode mode_closure;
+        let arg = Dummy {mode_closure; mode_arg; mode_ret; sort_arg } in
+        ty_ret, mode_closure, (lbl, arg) :: args
+    | (lbl, Omitted r) :: rest ->
+      let ty_ret, mode_ret, args = type_underscore_parameters closed_args rest in
+      ty_ret, mode_ret, (lbl, Omitted r) :: args
+    | (lbl, Arg r) :: rest ->
+      let ty_ret, mode_ret, args = type_underscore_parameters closed_args rest in
+      ty_ret, mode_ret, (lbl, Arg r) :: args
```

Comment:
You can change this to one case with an or-pattern:

```suggestion
    | (lbl, (Omitted _ | Arg _ as arg)) :: rest ->
      let ty_ret, mode_ret, args = type_underscore_parameters closed_args rest in
      ty_ret, mode_ret, (lbl, arg) :: args
```

`as` patterns can generalize types that aren't mentioned by the constructors that appear in them.